classlib/src/main/java/org/teavm/classlib/java/util/TTreeMap.java

```diff
@@ -753,36 +761,32 @@ public void remove() {
                 throw new TConcurrentModificationException();
             }
             if (last == null) {
-                throw new TNoSuchElementException();
-            }
-            var newRoot = owner.deleteNode(owner.root, last.getKey());
-            if (owner.root != newRoot) {
-                owner.root = newRoot;
-                var newPath = owner.pathToNext(last.getKey(), reverse);
-                System.arraycopy(newPath, 0, path, 0, newPath.length);
-                depth = newPath.length;
+                throw new IllegalStateException();
             }
+            owner.root = owner.deleteNode(owner.root, last.getKey());
```

Comment:
This still a performance issue, not as bad as previous one. The path will be rebuilt after each deletion, however tree is not re-balanced after each deletion. The path should be only re-built after re-balancing, right? What actually caused AIOOBE, did you find out?

Reply:
Depth should be recalculated. AIOOBE was due to wrong depth after the deletion. For recalculating depth you should get new path. That was the reason why I removed "if" check. And I suppose that this is unavoidable.

Reply:
Why should depth be recalculated? If three was not re-balanced, we have the same path leading to node being just removed. The next fetched node will always be searched from `depth - 1`, so it does not matter that `path[depth]` got changed. Even after re-balancing `path` length should be enough, since after deletion overall tree height can't grow.

Reply:
So I still *suspect* that there's an issue in the code that maintains tree height. If it's not the case and if you think that path recalculation is unavoidable, please, prove this. For example, you can describe sequence of steps that turn path into inconsistent state

Reply:
Depth is not related to tree height.
Let's consider next tree:
```
       4 6  9
0 2   5   8
 1      7
     3
```
As you can see, `3` has path that consists of `[3]` only, `4` has path that consists of `[7, 5, 4]`. I understand that after deletion of 3 tree will be rebalanced, but you definitely can invent tree with two branches, where after deletion depth will grow.

Reply:
> but you definitely can invent tree with two branches

Theorem, please!

My attempt for simple rotation:

Consider we have a subtree with root node X and balance factor +2. Let children of X are T1 and Y, children of Y are T2 and T3. height(Y) = height(T1) + 2, then height(T2) <= height(T1) + 1, height(T3) <= height(T1) + 1. height(X) = height(T1) + 3. 

After left turn we have Y1 as the root, where as X1 and T3 are its children, T1 and T2 are children of X. Then, height(X1)  = 1 + min(height(T1), height(T2)), but since height(T2) <= height(T1), min(height(T1), height(T2)) <= height(T1) + 2. height(Y1) = 1 + min(height(X), height(T3)), but min(height(X), height(T)) <= height(T1) + 2, so height(Y1) <= 3 + height(T1), Q.E.D.

Now you can write similar proof for double rotation.

Reply:
There is even worse issue in code basically.
I restored failure in the code.
Added next method:
```java
private boolean fillSet(Set<K> set, TreeNode<K, ?> node) {
            if (set.add(node.getKey())) {
                System.out.println("Duplicate " + node.getKey());
                return false;
            }
            return fillSet(set, node.left) && fillSet(set, node.right);
        }
```
and wrapped IAOOBE into next protective catch:
```java
int lastDepth = depth;
            last = node;
            TreeNode<K, V> down = node.down(reverse);
            if (down != null) {
                node = down;
                while (node != null) {
                    try {
                        path[depth++] = node;
                    } catch (ArrayIndexOutOfBoundsException e) {
                        System.out.println(owner.root.height + " " + lastDepth + " " + last.height + " " + owner.size() + " " + (depth - 1) + " " + Arrays.toString(path));
                        Set<K> s = new HashSet<>();
                        System.out.println(fillSet(s, owner.root));
                        //BTreePrinter.printNode(owner.root);
                        throw e;
                    }
                    node = node.forward(reverse);
                }
            }
```
Output:
```java
11 8 6 983 11 [60=60, 35=35, 22=22, 19=19, 140=140, 13=13, 122=122, 121=121, 13=13, 126=126, 122=122]
Duplicate 60
false
```
It seems that TreeMap allows duplicates in some cases.

Reply:
I tried to go other way. To print tree structure every iteration and find the place where bad structure happens. Unfortunately while System.out.println works in small cases, it for some reason doesn't work in iterations. It either ignores to go to output or goes to infinite loop where it shouldn't.

Reply:
Steps to reproduce: 
https://github.com/konsoletyper/teavm/pull/813/commits/23ea0b7379162a23c85f7a8b6c701521d0c9eb6b#diff-959d6fa49c32ea239164890ed5b82a3f3fc97b3c787ad33585c6bd1be475673aR731
check this commit. Run, see that test fails (and see the output). Uncomment commented line `//System.out.println(depth + " " + path.length);`. Run test. Check for infinite loop. Comment this "try-catch" structure instead. Test now fails, but all System.out.println was ignored and nothing goes to output.

Reply:
I tried to print it into console when launched test in IDEA during test `run` (not `debug`). No debugger, just System.out.println in this case.

Reply:
Is there any reason to not use debugger?

Reply:
I wanted to have log in the end and look through it. Set size is 1000, failing case is 22th. I know about conditional breakpoints, but it is too complicated to prepare all for debugging every time you changing something.

Reply:
Quick hint for printing a tree: there's a popular tool for rendering graphs, called graphviz. It has its language for graph representation. There's GraphBuilder class in TeaVM whish creates Graph, and Graph.toString produces graphviz format.

Reply:
Many thanks, will use it.
Still, this failure is not related to this copy-paste from StackOverflow, it's commented.

Reply:
In such cases solution is following: add verification method that throws exception when something gets inconsistent, then insert call to this method everywhere, then put exception breakpoint

Reply:
So I narrowed down problem. Looks like deletion can re-balance tree, and it's does not necessarily updates root. I'll investigate it further

Reply:
Ok, now I realized that it's impossible to incrementally update path (because it's not really whole path, but path containing only left-hand children), neither it's possible to even tell if path should be updated. But I can conclude, that idea of *not* keeping parent of each node was bad WRT performance. Even removal seems quite poor, requiring N*log(N) comparisons every time, even if it's performed from iterator. Perhaps, I'll refactor this later and simplify many things, i.e. size computation, navigation, iteration and so on.

Reply:
By the way, what WRT acronym mean you are often using?

Reply:
with respect to = regarding, concerning